### PR TITLE
Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56424 AutoDispatchBelowAutograd takes no arguments.
* #56423 s/AutoNonVariableTypeMode/AutoDispatchBelowAutograd/
* **#56422 Rename AutoNonVariableTypeMode to AutoDispatchBelowAutograd and add a warning.**
* #56434 Switch assertWarnsOnceRegex logic to check any instead of all.

Differential Revision: [D27866608](https://our.internmc.facebook.com/intern/diff/D27866608)

### CODE REVIEW

**Review by albanD — APPROVED** [review]
LGTM

File: aten/src/ATen/core/LegacyTypeDispatch.h
Comment (bhosmer) [inline aten/src/ATen/core/LegacyTypeDispatch.h:20]: We don't really use "below" as the name for this concept anywhere else, do we? Not that we should sweat this name too much, but we do use "after" in e.g. `DispatchKeySet::FullAfter`, so maybe `AutoDispatchAfterAutograd`?

**Review by bhosmer — APPROVED** [review]
(sorry, meant to approve - your call on "below" vs "after" naming suggestion ofc 😁 )

File: aten/src/ATen/core/LegacyTypeDispatch.h
Comment (ailzhang) [reply aten/src/ATen/core/LegacyTypeDispatch.h:20]: @bhosmer I named the other RAII guard `AutoDispatchBelowInplaceOrView` so we should either keep or change them both together I think ;)

Comment (bhosmer) [reply aten/src/ATen/core/LegacyTypeDispatch.h:20]: Ah you're absolutely right (and my bad for not raising the "below" thing there!) Anyway it's a very minor difference and the note explains exactly what the guard does in any case. Definitely not worth digging up the sidewalk.

**Pull request closed without merging**
